labs/intro-to-ada/chapters/solutions/standard_library.rst

```diff
@@ -0,0 +1,1031 @@
+       is
+          Q : Integer;
+          C : constant Item_C := Item_C (ID);
+       begin
+          case Trans is
+             when Bought =>
+                Q := Inv.List_Item (C).Stock_Quantity + Quantity;
+             when Sold =>
+                Q := Inv.List_Item (C).Stock_Quantity - Quantity;
+          end case;
+
+          if Q >= 0 then
+             Success := True;
+
+             Inv.List_Item (C).Stock_Quantity := Q;
+
+             Inv.List_Item (C).Stock_Amount :=
+               Amount (Q) * Inv.List_Item (C).Price;
+
+             Inv.List_Item (C).Trans_Quantity (Trans) :=
+               Inv.List_Item (C).Trans_Quantity (Trans) + Quantity;
+
+             Inv.List_Item (C).Trans_Amount (Trans) :=
+               Inv.List_Item (C).Trans_Amount (Trans) +
+               Amount (Quantity) * Inv.List_Item (C).Price;
+          else
+             Success := False;
+          end if;
+
+       end Set;
+
+       function Get (Inv   : Inventory;
+                     ID    : Item_ID) return String is
+         (To_String (Inv.List_Item (Item_C (ID)).Name));
+
+       function Get (Inv   : Inventory;
+                     ID    : Item_ID) return Item_Quantity is
+         (Inv.List_Item (Item_C (ID)).Stock_Quantity);
+
+       function Get (Inv   : Inventory;
+                     ID    : Item_ID) return Amount is
+         (Inv.List_Item (Item_C (ID)).Stock_Amount);
+
+       function Get (Inv   : Inventory;
+                     Trans : Transaction_Type;
+                     ID    : Item_ID) return Item_Quantity is
+         (Inv.List_Item (Item_C (ID)).Trans_Quantity (Trans));
+
+       function Get (Inv   : Inventory;
+                     Trans : Transaction_Type;
+                     ID    : Item_ID) return Amount is
+         (Inv.List_Item (Item_C (ID)).Trans_Amount (Trans));
+
+       function Get (Inv   : Inventory;
+                     Trans : Transaction_Type) return Amount
+       is
+          Total : Amount := 0.0;
+       begin
+          for C in Inv.List_Item.Iterate loop
+             Total := Total + Get (Inv, Trans, Item_ID (C));
+          end loop;
+
+          return Total;
+       end Get;
+
+       function Get (Inv   : Inventory) return Amount
+       is
+          Total : Amount := 0.0;
+       begin
+          for C in Inv.List_Item.Iterate loop
+             Total :=  + Get (Inv, Item_ID (C));
+          end loop;
+
+          return Total;
+       end Get;
+
+       procedure Display (Inv : Inventory)
+       is
+          package F_IO is new Ada.Text_IO.Decimal_IO (Amount);
+
+          use F_IO;
+       begin
+          for C in Inv.List_Item.Iterate loop
+             declare
+                I : constant Item_ID := Item_ID (C);
+             begin
+                Put_Line ("==== ITEM "
+                          & ": " & Get (Inv, I));
+                New_Line;
+                Put ("Price:     ");
+                Put (Amount'(Get (Inv, I)), 1, 2, 0);
```

Comment:
This is wrong. There is no accessor function for the 'price' field - this is returning the 'Stock_Amount' value.
I made some mistakes, and the use of Amount and Quantity to mean different things was confusing (should use "value" instead of "amount", because for "Overall" saying "Amount Bought" usually means quantity, not cost)

Reply:
Agreed, there's no _price_ field in that case. I've removed those lines. Also, I've replaced the wording ("amount"). Let me know what you think.

Reply:
Easier to understand now. thanks